### PR TITLE
Redux/redux thunk

### PR DESCRIPTION
**Pull request by EstebanIndiveri**
## Summary

Added redux-thunk, changed Book/Actions , Book/Reducers, Books interfaces , apply middleware

## Trello Card

[Trello Card](https://trello.com/c/wmlRYevd/28-n-92-redux-ii-middleware)

### CODE REVIEW

File: wbooks/src/redux/book/actions.ts [hunk 1]
```diff
@@ -1,8 +1,42 @@
-import { BookActionsTypes, BookInterface, BookType } from '@interfaces/book';
+import { BOOKS_MOCK } from '@constants/mockBooks';
+import { BookActionsTypes, BookInterface, BookState, BookType, BookPromise } from '@interfaces/book';
+import { Action, Dispatch } from 'redux';
+import { ThunkAction } from 'redux-thunk';
```
Comment (mattgle) [inline wbooks/src/redux/book/actions.ts:2]: these 2 should be at the top.

File: wbooks/src/redux/book/actions.ts [hunk 2]
```diff
@@ -1,8 +1,42 @@
...
+const fakeApiTimeOut = () => {
+  return new Promise<BookPromise>((resolve, reject) => {
+    setTimeout(() => {
+      const response = { ok: true, data: BOOKS_MOCK };
+      resolve(response);
+      const error = { ok: false, data: [] };
+      reject(error);
+    }, 300);
+  });
+};
```
Comment (mattgle) [inline wbooks/src/redux/book/actions.ts]: this should be at `services/books`

File: wbooks/src/redux/book/actions.ts [hunk 3]
```diff
@@ -1,8 +1,42 @@
...
+
+const setBooks = (books: BookInterface[]): BookActionsTypes => ({
+  type: BookType.BOOK_LIST_SUCCESS,
+  payload: books
+});
+
+const setBookError = (booksError: string): BookActionsTypes => ({
+  type: BookType.BOOK_LIST_ERROR,
+  booksError
+});
```
Comment (mattgle) [inline wbooks/src/redux/book/actions.ts]: This is not the way we manage actions. Read the card again pls.

File: wbooks/src/redux/book/reducer.ts [hunk 4]
```diff
@@ -1,22 +1,19 @@
...
   switch (action.type) {
     case BookType.BOOK_LIST:
+      return { bookLoading: true, books: [] };
+    case BookType.BOOK_LIST_SUCCESS:
       return { bookLoading: false, books: BOOKS_MOCK };
+    case BookType.BOOK_LIST_ERROR:
+      return { ...state, bookLoading: false, booksError: action.booksError };
     default:
       return state;
   }
```
Comment (mattgle) [inline wbooks/src/redux/book/reducer.ts]: Your reducer should look like this: 
```
const initialState: BooksState = {
  books: [],
  booksLoading: false,
  booksError: null
};

function booksReducer(state = initialState, action: BookAction): BooksState {
  switch (action.type) {
    case bookActions.GET_BOOKS:
      return { ...state, booksLoading: true };

    case bookActions.GET_BOOKS_SUCCESS:
      return { ...state, books: action.payload, booksLoading: false };

    case bookActions.GET_BOOKS_FAILURE:
      return { ...state, booksLoading: false, booksError: action.payload };

    default:
      return state;
  }
}
```

**Review by mattgle — CHANGES_REQUESTED** [review]

File: wbooks/src/services/book.tsx [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+export const fakeApiTimeOut = () => {
+  return new Promise<BookPromise>((resolve, reject) => {
+    setTimeout(() => {
+      const response = { ok: true, data: BOOKS_MOCK };
+      resolve(response);
+      const error = { ok: false, data: [], problem: CUSTOM_LIBRARY_ERROR };
+      reject(error);
+    }, 300);
+  });
+};
```
Comment (mattgle) [inline wbooks/src/services/book.tsx]:
```
export const fakeApiTimeOut = () =>
  new Promise(resolve =>
    setTimeout(() => {
      resolve({ ok: true, data: BOOKS_MOCK });
    }, 1000)
  );
```

**Review by mattgle — CHANGES_REQUESTED** [review]

File: wbooks/src/redux/book/actions.ts (hunk 1 above)
Comment (EstebanIndiveri) [reply wbooks/src/redux/book/actions.ts:2]: like this?

**Review by mattgle — APPROVED** [review]

**Pull request merged**